### PR TITLE
Feature result, closes #36 closes #37

### LINKED ISSUE #36: Election results should be calculated when Election ends
Whenever an election ends calculate the results and save them. Therefore, create Results table.

- [ ] Also no personal Voter data should be part of the results

### LINKED ISSUE #37: Election Results should be displayed
Election results should be displayed on the Election page or similar

### PR DESCRIPTION
**Pull request by danieliheonu**

### CODE REVIEW

File: app/models.py [hunk 1]
```diff
@@ -62,8 +62,12 @@ class Election(db.Model):
     number_of_voters = db.Column(db.String(), nullable=False)
     password = db.Column(db.String())
     candidates = db.relationship(
-        'Candidate', backref='election', lazy='dynamic')
-    votes = db.relationship('Vote', backref='election', lazy='dynamic')
+                        'Candidate', backref='election', lazy='dynamic', 
+                        passive_deletes=True, cascade="all, delete")
+    votes = db.relationship('Vote', backref='election', lazy='dynamic', 
+                            passive_deletes=True, cascade="all, delete")
+    result = db.relationship('Result', backref='candidate', lazy='dynamic',
```
Comment (LeanKhan) [inline app/models.py:69]: Why is the backref for results candidate

File: app/models.py [hunk 2]
```diff
@@ -79,8 +83,11 @@ class Candidate(db.Model):
     bio = db.Column(db.Text)
     position = db.Column(db.String(200))
     election_id = db.Column(db.Integer, db.ForeignKey(
-        'election.id'), nullable=False)
-    votes = db.relationship('Vote', backref='candidate', lazy='dynamic')
+        'election.id', ondelete='CASCADE'), nullable=False)
+    votes = db.relationship('Vote', backref='candidate', lazy='dynamic', 
+                            passive_deletes=True, cascade="all, delete")
+    result = db.relationship('Result', backref='candidate_result', lazy='dynamic',
```
Comment (LeanKhan) [inline app/models.py:89]: Backref should be "candidate" instead of "candidate_result"

File: app/models.py [hunk 3]
```diff
@@ -94,9 +101,9 @@ class Vote(db.Model):
     id = db.Column(db.Integer, primary_key=True)
     user_id = db.Column(db.Integer, db.ForeignKey('user.id'), nullable=True)
     election_id = db.Column(db.Integer, db.ForeignKey(
-        'election.id'), nullable=False)
+        'election.id', ondelete='CASCADE'), nullable=False)
```
Comment (LeanKhan) [inline app/models.py:104]: What does this do? Does it delete the Election or the Vote?

File: app/models.py [hunk 4]
```diff
@@ -106,6 +113,15 @@ def as_dict(self):
         return {item.name: getattr(self, item.name) for item in self.__table__.columns}
 
 
+class Result(db.Model):
+    id = db.Column(db.Integer, primary_key=True)
+    election_id = db.Column(db.ForeignKey(
+        'election.id', ondelete='CASCADE'), nullable=False)
+    candidate_id = db.Column(db.Integer, db.ForeignKey(
+        'candidate.id', ondelete='CASCADE'), nullable=False)
+    total_votes = db.Column(db.String())
```
Comment (LeanKhan) [inline app/models.py:122]: Does the type for `total_votes` being String affect anything?

File: app/routes.py [hunk 5]
```diff
@@ -99,15 +99,18 @@ def create_election():
...
+@app.route("/election/<link>", methods=['GET', 'POST'])
 def election(link):
     election = Election.query.filter_by(link=link).first()
     if election is None:
         flash(f'There is no such election', 'danger')
         return redirect(url_for('missing_route'))
-    if election.owner.id == current_user.id:
-        statuses = Status.query.all()
-    return render_template('election.html', title=election.name, election=election, statuses=statuses)
+    statuses = Status.query.all()
```
Comment (LeanKhan) [inline app/routes.py:109]: Are you fetching the statuses by default?

File: app/routes.py [hunk 6]
```diff
@@ -99,15 +99,18 @@ def create_election():
...
+    
+    form = ResultForm()
+    if form.validate_on_submit():
```
Comment (LeanKhan) [inline app/routes.py:112]: So anytime I make a POST request to the election page, is it the `ResultForm()` that will handle it?

Comment (LeanKhan) [reply app/routes.py:112]: I don't get what this part is for. So after a POST request, you want to redirect to the result page?

File: app/routes.py [hunk 7]
```diff
@@ -123,8 +126,9 @@ def delete_election(link):
     for c in election.candidates.all():
         db.session.delete(c)
     
-    for v in election.votes.all():
-        db.session.delete(v)
+    if election.votes.all() is not None:
```
Comment (LeanKhan) [inline app/routes.py:130]: I think this is Repeatitive. How about 
```suggestion
    votes = election.votes.all()
    if votes is not None:
        for v in votes:
        ...
```

So you won't have to make those calls twice

File: app/templates/election.html [hunk 8]
```diff
@@ -10,8 +10,12 @@ <h3 id="urlText" style="color: lime;background-color: gray;">{{ url_for('voting_
     <a href="{{ url_for('update_election', link=election.link) }}">edit</a>
     <a href="{{ url_for('delete_election', link=election.link) }}">delete</a>
     {% endif %}
+    
     <h4> Created by: {{ election.owner.email }} </h4>
+
+    {% if election.owner.id == current_user.id %}
```
Comment (LeanKhan) [inline app/templates/election.html:22]: why can only the owner of the election see this? sha no p

File: app/templates/election.html [hunk 9]
```diff
@@ -62,4 +66,10 @@ <h4>Candidates</h4>
         {{ election.description or 'No description' }}
     </p>
 </div>
+<p>
+    <form action="{{ url_for('result', link=election.link, slug=election.slug) }}"  method="GET">
```
Comment (LeanKhan) [inline app/templates/election.html:89]: what is this Form for? To take you to the Election Results page? Why not just a link? Better still, why can't we just render the results on this page?

File: app/models.py [hunk 10]
```diff
@@ -6,7 +6,7 @@
 class Type(db.Model):
 
     id = db.Column(db.Integer, primary_key=True)
-    name = db.Column(db.String())  # name of type
+    name = db.Column(db.String(), default='individual')  # name of type
```
Comment (LeanKhan) [inline app/models.py:9]: Why are we giving Type a default?
```suggestion
    name = db.Column(db.String())  # Type doesn't need a default. The default type is for **Users**
```

**Review by AnointingMax — APPROVED** [review]

File: app/models.py (hunk 3 above)
Comment (danieliheonu) [reply app/models.py:104]: It deletes the elections in all the tables

File: app/models.py (hunk 4 above)
Comment (danieliheonu) [reply app/models.py:122]: nope

File: app/routes.py (hunk 5 above)
Comment (danieliheonu) [reply app/routes.py:109]: it's the same as what was there before, I just made the changes in the template

File: app/routes.py (hunk 7 above)
Comment (danieliheonu) [reply app/routes.py:130]: it's basically the same thing but I will make the changes

File: app/templates/election.html (hunk 8 above)
Comment (danieliheonu) [reply app/templates/election.html:22]: I don't think the voters need to know when it was created

File: app/models.py (hunk 3 above)
Comment (LeanKhan) [reply app/models.py:104]: Why? Is that the desired behaviour? So if I delete one vote all the elections will be deleted? All? That's not what we want nau

If I delete a vote, it should only delete that vote.

File: app/routes.py (hunk 7 above)
Comment (LeanKhan) [reply app/routes.py:130]: same result yes, but not the same thing. You shouldnt be making two database calls when you only need one.

Comment (LeanKhan) [reply app/routes.py:130]: Database calls are expensive (memory and shii)

File: app/templates/election.html (hunk 8 above)
Comment (LeanKhan) [reply app/templates/election.html:22]: Why not?

File: app/models.py (hunk 3 above)
Comment (danieliheonu) [reply app/models.py:104]: No... This handles the deletion of an election. If you delete an election, it deletes everything related to it

Comment (danieliheonu) [reply app/models.py:104]: No... This handles the deletion of an election. If you delete an election, it deletes everything related to it

File: app/routes.py (hunk 7 above)
Comment (danieliheonu) [reply app/routes.py:130]: True. I changed it shah

Comment (danieliheonu) [reply app/routes.py:130]: True. I changed it shah

File: app/models.py (hunk 3 above)
Comment (danieliheonu) [reply app/models.py:104]: No... This handles the deletion of an election. If you delete an election, it deletes everything related to it

File: app/templates/election.html (hunk 8 above)
Comment (danieliheonu) [reply app/templates/election.html:22]: So you think the voters should know the date and time the election was created?

File: app/templates/election.html (hunk 9 above)
Comment (danieliheonu) [reply app/templates/election.html:89]: I will make a link button and the election can't be rendered until the end of the election

**Pull request merged**